Bug 1683058: Prevent sharing of configmap and secret data

sallyom opened the pull request:
This PR will ensure changes in ca bundle injector configmap and serving cert secret data are stomped on by the controller.

https://jira.coreos.com/browse/AUTH-204

mrogers950 commented on pkg/controller/servingcert/controller/secret_updating_controller_test.go:
```diff
@@ -368,6 +368,29 @@ func TestRequiresRegenerationServiceUIDMismatchBetaAnnotation(t *testing.T) {
 			expected: true,
 		},
+		{
+			name: "secret with added data",
```
I don't think this tests your related changes as it is just exercising the check against an already-queued secret to see if it matches up with its owner (yours is good, so returning "no regeneration" is to be expected). You should make a test in the e2e suite that creates a serving cert like normal and then modifies the secret, and then checks to see if it gets stomped on or not.

sallyom replied:
I updated the test, added e2e

mrogers950 commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -58,9 +58,16 @@ func (sc *serviceServingCertUpdateController) addSecret(obj metav1.Object) bool
+	// if the current data does not match or doesn't have a service name, check the old
 	// TODO drop this
-	return sc.addSecret(cur) || sc.addSecret(old)
+	if string(curSecret.Data[v1.TLSCertKey]) == string(oldSecret.Data[v1.TLSCertKey]) &&
```
This seems right to me but my brain wants to think about the inverse case (i.e., cur and old data do not match or cur does not have 2 items, so add old). Not sure if that is clearer to anyone else :)

stlaz replied:
+1, can't explain why my brain wants it that way.

sallyom replied:
lol ok, 2 against 1 :)

stlaz commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -58,9 +58,16 @@ func (sc *serviceServingCertUpdateController) addSecret(obj metav1.Object) bool
 	// TODO drop this
-	return sc.addSecret(cur) || sc.addSecret(old)
+	if string(curSecret.Data[v1.TLSCertKey]) == string(oldSecret.Data[v1.TLSCertKey]) &&
+		string(curSecret.Data[v1.TLSPrivateKeyKey]) == string(oldSecret.Data[v1.TLSPrivateKeyKey]) &&
```
`bytes.Equal()`

sallyom replied:
done, thanks

stlaz commented on test/e2e/e2e_test.go:
```diff
@@ -19,9 +20,10 @@ import (
-	serviceCAControllerNamespace = operatorclient.TargetNamespace
+	serviceCAOperatorNamespace = operatorclient.OperatorNamespace
+	serviceCAOperatorPodPrefix = operatorclient.OperatorNamespace // Same as operator namespace
+	//serviceCAControllerNamespace = operatorclient.TargetNamespace
```
Don't forget to remove this change :slightly_smiling_face:

sallyom replied:
yup, thanks, done

stlaz commented on test/e2e/e2e_test.go:
```diff
@@ -74,6 +76,19 @@ func createServingCertAnnotatedService(client *kubernetes.Clientset, secretName,
+		ObjectMeta: metav1.ObjectMeta{
+			Name: configMapName,
+			Annotations: map[string]string{
+				api.AlphaInjectCABundleAnnotationName: "true",
```
Shouldn't we be testing this with the current version now? Or maybe with both?

sallyom replied:
dunno, we'll clear that up b4 merging though

sallyom commented on pkg/controller/configmapcainjector/controller/configmap_injecting_controller.go:
```diff
@@ -46,20 +46,31 @@ func (ic *configMapCABundleInjectionController) Sync(obj v1.Object) error {
+	if data, ok := configMapCopy.Data[api.InjectionDataKey]; ok && data == ic.ca && len(configMapCopy.Data) == 1 {
+		return nil
+	}
 	if configMapCopy.Data == nil {
```
good point. thanks :) done

sallyom commented on pkg/controller/configmapcainjector/controller/configmap_injecting_controller.go:
```diff
@@ -46,20 +46,31 @@ func (ic *configMapCABundleInjectionController) Sync(obj v1.Object) error {
-	if data, ok := sharedConfigMap.Data[api.InjectionDataKey]; ok && data == ic.ca {
-		return nil
+	// ensure data of configmap
+	if err := ic.ensureConfigMapCABundleInjection(sharedConfigMap); err != nil {
```
good cleanup, thanks

sallyom commented on pkg/controller/configmapcainjector/controller/configmap_injecting_controller_test.go:
```diff
@@ -31,6 +31,9 @@ func validateActions(t *testing.T, expectedActionsNum int, actions []clienttesti
 		t.Error(diff.ObjectDiff(expected, actual))
 	}
+	if expectedDataLen := 1; len(actual.Data) != expectedDataLen {
+		t.Error(diff.ObjectDiff(expectedDataLen, len(actual.Data)))
```
doh, good point, thanks :) , done

sallyom commented:
@stlaz thanks for the review ;) I've updated and also added a test for updating serving-cert-secret in unit-tests and in extended.  Secret is regenerating for me now, in testing, when I give the existing secret bad data, so that's good.
@mrogers950 not sure if you're still seeing the issue with regeneration you were seeing last week? check to see if this fixes.

sallyom commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -313,6 +313,14 @@ func toRequiredSecret(dnsSuffix string, ca *crypto.CA, service *v1.Service, secr
 		secretCopy.Data = map[string][]byte{}
 	}
 
+	for k, _ := range secretCopy.Data {
```
k, done

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller_test.go:
```diff
@@ -60,6 +60,10 @@ func TestRequiresRegenerationServiceUIDMismatch(t *testing.T) {
 					},
 					OwnerReferences: []metav1.OwnerReference{ownerRef(&v1.Service{ObjectMeta: metav1.ObjectMeta{Name: "foo", UID: types.UID("uid-2")}})},
 				},
+				Data: map[string][]byte{
```
you are correct

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller_test.go:
```diff
@@ -79,6 +83,10 @@ func TestRequiresRegenerationServiceUIDMismatch(t *testing.T) {
 					},
 					OwnerReferences: []metav1.OwnerReference{ownerRef(&v1.Service{ObjectMeta: metav1.ObjectMeta{Name: "foo", UID: types.UID("uid-1")}})},
 				},
+				Data: map[string][]byte{
```
yup, thanks

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -114,6 +123,20 @@ func (sc *serviceServingCertUpdateController) requiresRegeneration(secret *v1.Se
 		return true, sharedService
 	}
 
```
: ) looking at it now, will update

sallyom replied:
updated

mrogers950 commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -309,9 +309,11 @@ func toRequiredSecret(dnsSuffix string, ca *crypto.CA, service *v1.Service, secr
-	}
+	// let garbage collector cleanup map allocation, for simplicity
+	secretCopy.Data = map[string][]byte{
+    v1.TLSCertKey: certBytes,
```
A bad indent here

sallyom replied:
fixed : )

mrogers950 commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -114,6 +123,20 @@ func (sc *serviceServingCertUpdateController) requiresRegeneration(secret *v1.Se
+			}
+		}
+	}
+	if len(secret.Data) != 2 {
```
With the change to updateSecret() I don't think we can hit these conditions. updateSecret() makes it so that on update if there is an additional data item or someone fudges with the crt/key, the existing data is kept, so we make them immutable on those conditions.  What you added here wants to regenerate on those conditions. Thinking about regeneration vs. protection of the data, it would be better to have protection for the secret update controller and leave the intentional regeneration to the secret creating controller when you take the action of deleting the secret.

sallyom replied:
If I edit an existing secret with bad data(anything different), updateSecret() does not result in that change being set back to the old(good) data.  (I'm probably missing something, but the change in updateSecret() is not enough - regenerating is the only way I've found to get back the secret w/ good data).

sallyom replied:
refactored this, checking/replacing/regenerating as needed - my this PR is getting messy

mrogers950 commented on test/e2e/e2e_test.go:
```diff
@@ -87,6 +102,88 @@ func pollForServiceServingSecret(client *kubernetes.Clientset, secretName, names
+	})
+}
+
+func editServiceServingSecretData(client *kubernetes.Clientset, secretName, namespace string) error {
```
This should be addToServiceServingSecretData() and we can make editServiceServingSecretData() a function for another test where we fiddle with the existing data and ensure that the old secret persists.

sallyom replied:
changed it

stlaz commented on pkg/controller/configmapcainjector/controller/configmap_injecting_controller.go:
```diff
@@ -46,20 +46,21 @@ func (ic *configMapCABundleInjectionController) Sync(obj v1.Object) error {
-	configMapCopy.Data[api.InjectionDataKey] = ic.ca
-
-	_, err := ic.configMapClient.ConfigMaps(configMapCopy.Namespace).Update(configMapCopy)
+	configMapCopy.Data = map[string]string{api.InjectionDataKey: ic.ca}
```
Do the deep copy before this line instead, you're not modifying anything in the code above.

sallyom replied:
done

stlaz commented on pkg/controller/servingcert/controller/secret_creating_controller.go:
```diff
@@ -299,18 +302,23 @@ func toRequiredSecret(dnsSuffix string, ca *crypto.CA, service *v1.Service, secr
-	if secretCopy.Data == nil {
-		secretCopy.Data = map[string][]byte{}
+	// let garbage collector cleanup map allocation, for simplicity
+	secretCopy.Data = map[string][]byte{
```
lines 326, 327 should be removed now

sallyom replied:
yup, : ) fixed

sallyom commented:
/retest

sallyom commented:
/test e2e-aws

sallyom commented:
/test e2e-aws

mrogers950 commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +70,105 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+		return true
 	}
 
 	if time.Now().Add(sc.minTimeLeftForCert).After(expiry) {
```
Use the sc.minTimeLeftForCert or the minTimeLeft argument (using the struct member instead of an extra argument is fine)

sallyom commented:
/test e2e-aws

mrogers950 commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +70,105 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+		}
+		// check that we only have the data we care about
+		// if new data does not match, regenerate it
+		if !bytes.Equal(secretCopy.Data[v1.TLSCertKey], data[v1.TLSCertKey]) ||
```
I might be missing something but it seems since you're creating data new each time, it will contain a new cert+key and never match the existing secret, so this path will always lead to regeneration. 

In thinking through the different ways to modify the secret data and what our reaction to it in this else clause should be:
A. Someone adds in an extra data item. We want to get rid of this but keep the same tls.crt and tls.key, so for this we can have a standalone check for len(secretCopy.Data) > 2. If there is more than 2 items, then we make a new data map with the existing tls.crt and tls.key and update.
B. Someone edits tls.crt or tls.key data, replacing them with ascii art. Here we could x509 decode secretCopy data to verify that. If it's been munged then we update with the new cert data. If it's OK then we assume it's the serving cert. If someone replaced the cert data with a legitimate cert unless we had some kind of cache of the current certs issued by the serving CA (or we did an actual verify against the CA) we could not confirm that it belongs.

sallyom replied:
makes sense, I'll update

sallyom replied:
updated

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+		if err := toRequiredSecret(sc.dnsSuffix, sc.ca, service, secretCopy); err != nil {
+			return false, err
+		}
+		update = true
```
right, done.

sallyom replied:
it's good to check, though?, in case someone replaced the cert data with nonsense
(this won't prevent the case where someone for some reason replaces cert data with another valid-ish x509.Cert (ie, it's at least a parseable cert) but will prevent someone from replacing cert data with acsii art)

sallyom replied:
yes, right  I should have, I separated toRequiredSecret from ensureSecretData now, look ok?

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+	// 2. ensure cert data is at least a parseable certificate, if not replace with valid data
+	// This does not ensure that somebody didn't swap out secret data for another valid cert.
+	_, ok := secretCopy.Data[v1.TLSCertKey]
+	_, ok2 := secretCopy.Data[v1.TLSPrivateKeyKey]
```
thanks, done : )

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+	// 1. ensure only the 2 data keys we want exist, TLSCertKey and TLSPrivateKeyKey
+	// 2. ensure cert data is at least a parseable certificate, if not replace with valid data
+	// This does not ensure that somebody didn't swap out secret data for another valid cert.
+	_, ok := secretCopy.Data[v1.TLSCertKey]
```
oh right, duh thanks! : ) done

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+
+func (sc *serviceServingCertUpdateController) regenerateSecondCheck(service *v1.Service, secretCopy *v1.Secret) (bool, error) {
+	update := false
+	// If not regenerating, perform checks here to
```
ok, done

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+	return false
+}
+
+func (sc *serviceServingCertUpdateController) regenerateSecondCheck(service *v1.Service, secretCopy *v1.Secret) (bool, error) {
```
yes, you're right, working on that now...

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+	return service
+}
 
+func (sc *serviceServingCertUpdateController) requiresRegenerationFirstCheck(service *v1.Service, secret *v1.Secret, minTimeLeft time.Duration) bool {
```
done : )

sallyom commented:
@mrogers, @Standa, I've added unit-tests for cert replacing/regeneration/removing extra data fields:
For cases where expectedRegeneration = False, I've added: 
* all good, parseable cert, tlsCert/Key data fields - expectReplace = False, expectUpdate = False
* not parseable x509 cert - expectReplace = True, expectUpdate = True
* good cert data, but extra data field - expectReplace = False, expectUpdate = True
* bad data field, no tlsCert/tlsKey field: expectReplace = True, expectUpdate = True

stlaz commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+		if err != nil {
+			return err
+		}
+		if replace {
```
This should have stayed inside `ensureSecretData()`. Its presence there should also explain your question in https://github.com/openshift/service-ca-operator/pull/26/files#r270982624 (`toRequiredSecret()` generates valid data).

sallyom replied:
right, but if we didn't call toRequiredSecret, we should check the cert is a cert. Is toRequiredSecret ok separate from ensureSecretData? It makes it convenient to unit-test, to see whether we are correctly replacing the data or not.

stlaz commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+		// Regenerate the secret
+		glog.Infof("Error decoding cert bytes %s from secret: %s namespace: %s, replacing cert", v1.TLSCertKey, secretCopy.Name, secretCopy.Namespace)
+		return true, true
+	} else {
```
No need for an `else` as you're returning above, just do it outside any condition scope.

sallyom replied:
ugh again :(  'elses' fixed.

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller_test.go:
```diff
@@ -158,9 +178,38 @@ func TestRequiresRegenerationServiceUIDMismatch(t *testing.T) {
 			expected: false,
 		},
+		{
+			name: "bad data",
```
done

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,134 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+		if err := toRequiredSecret(sc.dnsSuffix, sc.ca, service, secretCopy); err != nil {
+			return err
+		}
+		update = true
```
@stlaz you meant I should update and return here? I do see that.. I'm sorry, this PR is getting difficult to manage.. must merge this today : (

sallyom replied:
nm I see now, it was above.. I commented there.

stlaz commented on pkg/controller/servingcert/controller/secret_updating_controller.go:
```diff
@@ -69,70 +72,142 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+	// 2. ensure cert data is at least a parseable certificate, if not replace with valid data
+	// This does not ensure that somebody didn't swap out secret data for another valid cert.
+	var err error
+	update, err = sc.ensureSecretData(service, secretCopy)
```
Nit: you can scrape `update` and `err` definition and declaration above now and only have `update, err := sc.ensureSecretData(service, secretCopy)`

sallyom replied:
yes, you are correct : ) done

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller_test.go:
```diff
@@ -179,21 +302,41 @@ func TestRequiresRegenerationServiceUIDMismatch(t *testing.T) {
+				if tc.serviceIsValid != isValid {
+					t.Errorf("validService result: %v unexpected", isValid)
+				}
+				if tc.serviceIsValid {
```
I removed checks for ensureSecretData here altogether.
1) validation prevents one from updating to a secret without the 2 required data keys (tlscert,key)
2) I added e2e tests to check that extra data is removed, secret w/ invalid cert data is regenerated.

sallyom commented on pkg/controller/servingcert/controller/secret_updating_controller_test.go:
```diff
@@ -179,21 +302,41 @@ func TestRequiresRegenerationServiceUIDMismatch(t *testing.T) {
+						t.Errorf("%s: expectedRegen %v, got %v", tc.name, tc.expectedRegen, actualRegen)
+					}
+					if !tc.expectedRegen {
+						update, actualReplace := c.ensureSecretData(service, tc.secret)
```
i've updated unit tests, no longer required

stlaz commented:
/lgtm
